x/stake/endblock.go

```diff
@@ -10,9 +10,15 @@ import (
 )
 
 func EndBlocker(ctx sdk.Context, k keeper.Keeper) (validatorUpdates []abci.ValidatorUpdate, completedUbds []types.UnbondingDelegation) {
+	// only change validator set in breath block after BEP159
 	var events sdk.Events
-	_, validatorUpdates, completedUbds, _, events = handleValidatorAndDelegations(ctx, k)
 	var csEvents sdk.Events
+	if !sdk.IsUpgrade(sdk.BEP159) {
+		_, validatorUpdates, completedUbds, _, events = handleValidatorAndDelegations(ctx, k)
+	} else {
+		k.DistributeInBlock(ctx, types.ChainIDForBeaconChain)
+		validatorUpdates = k.PopPendingABCIValidatorUpdate(ctx)
```

Comment:
wondering, if a newly created validator, we do undelegate to below minimal self delegation, what will happen?

Reply:
Since we don't do unbond in `undelegate`, it's fine to just jail it even it's not boned.

It reminds me of that we should add additional condition(`&& ctx.SideChainId() == "" && validator.IsBonded()`) for new update validator code.

```
// x/stake/keeper/delegation.go L581-590

	if validator.IsSelfDelegator(delegation.DelegatorAddr) && !validator.Jailed &&
		validator.TokensFromShares(delegation.Shares).RawInt() < k.MinSelfDelegation(ctx) {
		k.jailValidator(ctx, validator)
		k.OnSelfDelDropBelowMin(ctx, valAddr)
		if sdk.IsUpgrade(sdk.BEP159) && ctx.SideChainId() == "" && validator.IsBonded() {
			k.AddPendingABCIValidatorUpdate(ctx, []abci.ValidatorUpdate{validator.ABCIValidatorUpdateZero()})
			k.DeleteLastValidatorPower(ctx, validator.OperatorAddr)
		}
		validator = k.mustGetValidator(ctx, validator.OperatorAddr)
	}
```